cuqipy_fenics/testproblem.py

```diff
@@ -147,3 +155,275 @@ def form(m,u,p):
+            form,
+            mesh,
+            parameter_function_space=V,
+            solution_function_space=V,
+            dirichlet_bcs=dirichlet_bcs,
+            adjoint_dirichlet_bcs=adjoint_dirichlet_bcs)
+
+        # Create the domain geometry
+        G_FEM = FEniCSContinuous(V)
+        if field_params is None:
+            field_params = {}
+        if field_type is None:
+            G_domain = G_FEM
+        elif field_type == 'KL':
+            if field_params == {}:
+                field_params = {'length_scale': 0.1, 'num_terms': 32}
+            G_domain = MaternExpansion(G_FEM, **field_params)
+        else:
+            raise ValueError('Unknown field type.')
+
+        # Create the range geometry
+        G_range = FEniCSContinuous(V)
+
+        # Create the forward model
+        A = PDEModel(PDE, domain_geometry=G_domain, range_geometry=G_range)
+
+        # Create the prior
+        if prior is None:
+            prior = Gaussian(np.zeros(A.domain_dim), 1,
+                             geometry=G_domain, name='x')
+        elif prior.name != 'x':
+            raise ValueError('Prior name is expected to be "x".')
+
+        # Set up the exact solution
+        if exactSolution is None and field_type == "KL":
+            np.random.seed(15)
+            exactSolution = np.random.randn(G_domain.par_dim)
+        elif exactSolution is None:
+            def exactSolution(x): return 1.5 + 0.5 * \
+                np.sin(2*np.pi*x[0])*np.sin(2*np.pi*x[1])
+
+        if isinstance(exactSolution, np.ndarray):
+            exactSolution = cuqi.array.CUQIarray(
+                exactSolution,
+                is_par=True,
+                geometry=G_domain)
+
+        elif callable(exactSolution):
+            exactSolution_expr = to_dolfin_expression(exactSolution)
+            exactSolution_func = dl.interpolate(exactSolution_expr, V)
+            exactSolution = cuqi.array.CUQIarray(
+                exactSolution_func,
+                is_par=False,
+                geometry=G_domain)
+
+        else:
+            raise ValueError(
+                'exactSolution should be a numpy array, a function or None.')
+        print(exactSolution.__class__)
+
+        # Create the exact data
+        exact_data = A(exactSolution)
+        if not isinstance(exact_data, cuqi.array.CUQIarray):
+            exact_data = cuqi.array.CUQIarray(
+                exact_data, is_par=True, geometry=G_range)
+
+        # Create the data distribution and the noisy data
+        noise = np.random.randn(len(exact_data))
+        if relative_noise_std is None:
+            relative_noise_std = 0.01
+        noise_std = relative_noise_std * \
+            np.linalg.norm(exact_data)/np.linalg.norm(noise)
+        noise = noise_std*noise
+        data = exact_data + noise
+
+        y = Gaussian(mean=A(prior), cov=noise_std**2, geometry=G_range)
+
+        # Create the Bayesian problem
+        super().__init__(y, prior, y=data)
```

Comment:
I think this looks a bit confusing to the untrained. Perhaps stick with only x and y or "data_dist" and "prior".

Could be something like 

```python
 data_dist = Gaussian(mean=A(prior), cov=noise_std**2, geometry=G_range, name="y")
super().__init__(data_dist, prior)
super().set_data(data)
```

Reply:
Ah yes :)